synapse/http/server.py

```diff
@@ -515,52 +515,62 @@ def __init__(
     ):
         self._request = request
         self._iterator = iterator
+        self._paused = False
 
-    def start(self) -> None:
-        self._request.registerProducer(self, False)
+        # Register the producer and start producing data.
+        self._request.registerProducer(self, True)
+        self.resumeProducing()
 
     def _send_data(self, data: List[bytes]) -> None:
         """
-        Send a list of strings as a response to the request.
+        Send a list of bytes as a chunk of a response.
         """
         if not data:
             return
         self._request.write(b"".join(data))
 
+    def pauseProducing(self) -> None:
+        self._paused = True
+
     def resumeProducing(self) -> None:
         # We've stopped producing in the meantime (note that this might be
         # re-entrant after calling write).
         if not self._request:
             return
 
-        # Get the next chunk and write it to the request.
-        #
-        # The output of the JSON encoder is coalesced until min_chunk_size is
-        # reached. (This is because JSON encoders produce a very small output
-        # per iteration.)
-        #
-        # Note that buffer stores a list of bytes (instead of appending to
-        # bytes) to hopefully avoid many allocations.
-        buffer = []
-        buffered_bytes = 0
-        while buffered_bytes < self.min_chunk_size:
-            try:
-                data = next(self._iterator)
-                buffer.append(data)
-                buffered_bytes += len(data)
-            except StopIteration:
-                # The entire JSON object has been serialized, write any
-                # remaining data, finalize the producer and the request, and
-                # clean-up any references.
-                self._send_data(buffer)
-                self._request.unregisterProducer()
-                self._request.finish()
-                self.stopProducing()
-                return
-
-        self._send_data(buffer)
+        self._paused = False
+
+        # Write until there's backpressure telling us to stop.
+        while not self._paused:
+            # Get the next chunk and write it to the request.
+            #
+            # The output of the JSON encoder is coalesced until min_chunk_size is
+            # reached. (This is because JSON encoders produce a very small output
+            # per iteration.)
+            #
+            # Note that buffer stores a list of bytes (instead of appending to
+            # bytes) to hopefully avoid many allocations.
+            buffer = []
+            buffered_bytes = 0
+            while buffered_bytes < self.min_chunk_size:
+                try:
+                    data = next(self._iterator)
+                    buffer.append(data)
+                    buffered_bytes += len(data)
```

Comment:
Looking at the code this is basically what [`iocpreactor.abstract.FileHandle`](https://github.com/twisted/twisted/blob/twisted-19.2.0/src/twisted/internet/iocpreactor/abstract.py#L271-L279) is doing, so I think we can just skip all this and call `self._request.write(data)` for each chunk?

Reply:
The issue is that the [`http.Request`](https://github.com/twisted/twisted/blob/1b439fbabc584f56c7bf6abf754395ba1ed3800b/src/twisted/web/http.py#L1173-L1174) calls [`toChunk`](https://github.com/twisted/twisted/blob/1b439fbabc584f56c7bf6abf754395ba1ed3800b/src/twisted/web/http.py#L329-L337) so each time we call `write` it will get turned into an HTTP chunk.

The goal of this code is to avoid `b"{"` being emitted from the `JSONProducer` becoming `1\r\n{\r\n`, which is a large explosion in the streamed bytes.

Reply:
Oh gah I hate that `http.Request` gets in the way :/

Reply:
Can you add a comment to that effect please?

Reply:
I expanded the comment above this section in 2afe30189267532baadd95023d7cb84b8bab23e5.